instrumentation-annotation-support/src/main/java/io/opentelemetry/instrumentation/annotation/support/AttributeBindingFactory.java

```diff
@@ -0,0 +1,341 @@
+            }
+          };
+      setter.setAttribute(key, wrapper);
+    };
+  }
+
+  private static AttributeBinding doubleArrayBinding(String name) {
+    AttributeKey<List<Double>> key = AttributeKey.doubleArrayKey(name);
+    return (setter, arg) -> {
+      double[] array = (double[]) arg;
+      List<Double> wrapper =
+          new AbstractList<Double>() {
+            @Override
+            public Double get(int index) {
+              return array[index];
+            }
+
+            @Override
+            public int size() {
+              return array.length;
+            }
+          };
+      setter.setAttribute(key, wrapper);
+    };
+  }
+
+  private static AttributeBinding booleanArrayBinding(String name) {
+    AttributeKey<List<Boolean>> key = AttributeKey.booleanArrayKey(name);
+    return (setter, arg) -> {
+      boolean[] array = (boolean[]) arg;
+      List<Boolean> wrapper =
+          new AbstractList<Boolean>() {
+            @Override
+            public Boolean get(int index) {
+              return array[index];
+            }
+
+            @Override
+            public int size() {
+              return array.length;
+            }
+          };
+      setter.setAttribute(key, wrapper);
+    };
+  }
+
+  private static AttributeBinding defaultArrayBinding(String name) {
+    AttributeKey<List<String>> key = AttributeKey.stringArrayKey(name);
+    return (setter, arg) -> {
+      Object[] array = (Object[]) arg;
+      List<String> wrapper =
+          new AbstractList<String>() {
+            @Override
+            public String get(int index) {
+              Object value = array[index];
+              return value != null ? value.toString() : null;
+            }
+
+            @Override
+            public int size() {
+              return array.length;
+            }
+          };
+      setter.setAttribute(key, wrapper);
+    };
+  }
+
+  private static <T, U> AttributeBinding mappedListBinding(
+      Class<T> fromClass, AttributeKey<List<U>> key, Function<T, U> mapping) {
+    return (setter, arg) -> {
+      @SuppressWarnings("unchecked")
+      List<T> list = (List<T>) arg;
+      List<U> wrapper =
+          new AbstractList<U>() {
+            @Override
+            public U get(int index) {
+              T value = list.get(index);
+              return value != null ? mapping.apply(value) : null;
+            }
+
+            @Override
+            public int size() {
+              return list.size();
+            }
+          };
+      setter.setAttribute(key, wrapper);
+    };
+  }
+
+  private static AttributeBinding defaultListBinding(String name) {
+    AttributeKey<List<String>> key = AttributeKey.stringArrayKey(name);
+    return mappedListBinding(Object.class, key, Object::toString);
+  }
+
+  private static AttributeBinding defaultBinding(String name) {
+    AttributeKey<String> key = AttributeKey.stringKey(name);
+    return (setter, arg) -> setter.setAttribute(key, arg.toString());
```

Comment:
I have the consuming class perform the `null` check so that I could keep the `AttributeBinding` implementations/lambdas simple:  https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/56e72c5ac69700ce9149daeba48a718e8ae26eca/instrumentation-annotation-support/src/main/java/io/opentelemetry/instrumentation/annotation/support/BaseAttributeBinder.java#L100

I could move the `null` check to the `AttributeBinding` itself if that would be preferred.

Reply:
Oh I missed that, makes sense where you have it 👍